File: yandex-game/generate-audio.py
# Генерация звуковых файлов для Cars RNG

import wave
import struct
import math
import os

def create_wav(filename, samples, sample_rate=44100, channels=1):
    """Создание WAV файла из массива сэмплов"""
    with wave.open(filename, 'w') as wav_file:
        wav_file.setnchannels(channels)
        wav_file.setsampwidth(2)  # 16-bit
        wav_file.setframerate(sample_rate)

        for sample in samples:
            # Конвертация в 16-bit integer
            value = int(max(-1, min(1, sample)) * 32767)
            packed = struct.pack('<h', value)
            wav_file.writeframes(packed)

def generate_click():
    """Генерация звука клика"""
    sample_rate = 44100
    duration = 0.1
    samples = int(sample_rate * duration)

    buffer = []
    for i in range(samples):
        t = i / sample_rate
        envelope = math.exp(-t * 50)
        sample = math.sin(2 * math.pi * 800 * t) * envelope * 0.3
        buffer.append(sample)

    return buffer

def generate_button():
    """Генерация звука кнопки"""
    sample_rate = 44100
    duration = 0.08
    samples = int(sample_rate * duration)

    buffer = []
    for i in range(samples):
        t = i / sample_rate
        envelope = math.exp(-t * 40)
        sample = math.sin(2 * math.pi * 600 * t) * envelope * 0.25
        buffer.append(sample)

    return buffer

def generate_spin():
    """Генерация звука спина"""
    sample_rate = 44100
    duration = 1.5
    samples = int(sample_rate * duration)

    buffer = []
    for i in range(samples):
        t = i / sample_rate
        frequency = 200 + t * 400
        envelope = math.sin(math.pi * t / duration) * 0.4
        sample = math.sin(2 * math.pi * frequency * t) * envelope
        buffer.append(sample)

    return buffer

def generate_win():
    """Генерация звука выигрыша"""
    sample_rate = 44100
    duration = 0.8
    samples = int(sample_rate * duration)

    frequencies = [523, 659, 784, 1047]  # C5, E5, G5, C6
    buffer = []

    for i in range(samples):
        t = i / sample_rate
        note = int(t * 8)
        freq = frequencies[note % len(frequencies)]
        envelope = math.exp(-t * 3) * 0.5
        sample = math.sin(2 * math.pi * freq * t) * envelope
        buffer.append(sample)

    return buffer

def generate_case_open():
    """Генерация звука открытия кейса"""
    sample_rate = 44100
    duration = 0.6
    samples = int(sample_rate * duration)

    buffer = []
    for i in range(samples):
        t = i / sample_rate
        frequency = 1200 * math.exp(-t * 4)
        envelope = math.exp(-t * 5) * 0.4
        sample = math.sin(2 * math.pi * frequency * t) * envelope
        buffer.append(sample)

    return buffer

def generate_background_music():
    """Генерация фоновой музыки"""
    sample_rate = 44100
    duration = 8  # 8 секунд loop
    samples = int(sample_rate * duration)

    bass_notes = [130.81, 196.00, 220.00, 174.61]  # C3, G3, A3, F3
    melody_notes = [523.25, 587.33, 659.25, 783.99]  # C5, D5, E5, G5

    buffer = []
    for i in range(samples):
        t = i / sample_rate

        # Бас
        bass_index = int(t * 2) % len(bass_notes)
        bass_freq = bass_notes[bass_index]
        bass_envelope = math.sin(math.pi * (t % 0.5) / 0.5) * 0.15
        bass = math.sin(2 * math.pi * bass_freq * t) * bass_envelope

        # Мелодия
        melody_index = int(t) % len(melody_notes)
        melody_freq = melody_notes[melody_index]
        melody_envelope = math.sin(math.pi * (t % 1) / 1) * 0.1
        melody = math.sin(2 * math.pi * melody_freq * t) * melody_envelope

        # Микс
        sample = bass * 0.6 + melody * 0.4
        buffer.append(sample)

    return buffer

def main():
    print("🎵 Генерация звуковых файлов для Cars RNG...\n")

    # Создание директорий
    base_dir = os.path.join("public", "assets", "audio")
    music_dir = os.path.join(base_dir, "music")
    sounds_dir = os.path.join(base_dir, "sounds")

    os.makedirs(music_dir, exist_ok=True)
    os.makedirs(sounds_dir, exist_ok=True)

    # Генерация звуков
    sounds = {
        os.path.join(sounds_dir, "click.wav"): generate_click(),
        os.path.join(sounds_dir, "button.wav"): generate_button(),
        os.path.join(sounds_dir, "spin.wav"): generate_spin(),
        os.path.join(sounds_dir, "win.wav"): generate_win(),
        os.path.join(sounds_dir, "case-open.wav"): generate_case_open(),
        os.path.join(music_dir, "background-music.wav"): generate_background_music(),
    }

    # Сохранение файлов
    for filename, samples in sounds.items():
        create_wav(filename, samples)
        print(f"✅ Создан: {filename}")

    print("\n🎉 Все звуки успешно сгенерированы!")
    print(f"📁 Папка: {base_dir}")

if __name__ == "__main__":
    main()

File: yandex-game/generate-audio.js
#!/usr/bin/env node

/**
 * Скрипт для генерации звуковых файлов для игры Cars RNG
 * Использует Web Audio API через node-web-audio-api
 */

const fs = require('fs');
const path = require('path');

// Простая генерация WAV файлов без внешних зависимостей
class SimpleAudioGenerator {
  constructor(sampleRate = 44100) {
    this.sampleRate = sampleRate;
  }

  // Генерация звука клика
  generateClick() {
    const duration = 0.1;
    const samples = Math.floor(this.sampleRate * duration);
    const buffer = new Float32Array(samples);

    for (let i = 0; i < samples; i++) {
      const t = i / this.sampleRate;
      const envelope = Math.exp(-t * 50);
      buffer[i] = Math.sin(2 * Math.PI * 800 * t) * envelope * 0.3;
    }

    return buffer;
  }

  // Генерация звука кнопки
  generateButton() {
    const duration = 0.08;
    const samples = Math.floor(this.sampleRate * duration);
    const buffer = new Float32Array(samples);

    for (let i = 0; i < samples; i++) {
      const t = i / this.sampleRate;
      const envelope = Math.exp(-t * 40);
      buffer[i] = Math.sin(2 * Math.PI * 600 * t) * envelope * 0.25;
    }

    return buffer;
  }

  // Генерация звука спина
  generateSpin() {
    const duration = 1.5;
    const samples = Math.floor(this.sampleRate * duration);
    const buffer = new Float32Array(samples);

    for (let i = 0; i < samples; i++) {
      const t = i / this.sampleRate;
      const frequency = 200 + t * 400;
      const envelope = Math.sin(Math.PI * t / duration) * 0.4;
      buffer[i] = Math.sin(2 * Math.PI * frequency * t) * envelope;
    }

    return buffer;
  }

  // Генерация звука выигрыша
  generateWin() {
    const duration = 0.8;
    const samples = Math.floor(this.sampleRate * duration);
    const buffer = new Float32Array(samples);

    for (let i = 0; i < samples; i++) {
      const t = i / this.sampleRate;
      const note = Math.floor(t * 8);
      const frequencies = [523, 659, 784, 1047];
      const freq = frequencies[note % frequencies.length];
      const envelope = Math.exp(-t * 3) * 0.5;
      buffer[i] = Math.sin(2 * Math.PI * freq * t) * envelope;
    }

    return buffer;
  }

  // Генерация звука открытия кейса
  generateCaseOpen() {
    const duration = 0.6;
    const samples = Math.floor(this.sampleRate * duration);
    const buffer = new Float32Array(samples);

    for (let i = 0; i < samples; i++) {
      const t = i / this.sampleRate;
      const frequency = 1200 * Math.exp(-t * 4);
      const envelope = Math.exp(-t * 5) * 0.4;
      buffer[i] = Math.sin(2 * Math.PI * frequency * t) * envelope;
    }

    return buffer;
  }

  // Генерация фоновой музыки
  generateBackgroundMusic() {
    const duration = 8;
    const samples = Math.floor(this.sampleRate * duration);
    const bufferL = new Float32Array(samples);
    const bufferR = new Float32Array(samples);

    const bassNotes = [130.81, 196.00, 220.00, 174.61];
    const melodyNotes = [523.25, 587.33, 659.25, 783.99];

    for (let i = 0; i < samples; i++) {
      const t = i / this.sampleRate;

      const bassIndex = Math.floor(t * 2) % bassNotes.length;
      const bassFreq = bassNotes[bassIndex];
      const bassEnvelope = Math.sin(Math.PI * (t % 0.5) / 0.5) * 0.15;
      bufferL[i] = Math.sin(2 * Math.PI * bassFreq * t) * bassEnvelope;

      const melodyIndex = Math.floor(t) % melodyNotes.length;
      const melodyFreq = melodyNotes[melodyIndex];
      const melodyEnvelope = Math.sin(Math.PI * (t % 1) / 1) * 0.1;
      bufferR[i] = Math.sin(2 * Math.PI * melodyFreq * t) * melodyEnvelope;

      bufferL[i] += bufferR[i] * 0.5;
      bufferR[i] += bufferL[i] * 0.3;
    }

    return { left: bufferL, right: bufferR };
  }

  // Конвертация в WAV формат
  toWav(buffer, stereo = false) {
    const numberOfChannels = stereo ? 2 : 1;
    const length = (stereo ? buffer.left.length : buffer.length) * numberOfChannels * 2;
    const arrayBuffer = Buffer.alloc(44 + length);

    // WAV заголовок
    arrayBuffer.write('RIFF', 0);
    arrayBuffer.writeUInt32LE(36 + length, 4);
    arrayBuffer.write('WAVE', 8);
    arrayBuffer.write('fmt ', 12);
    arrayBuffer.writeUInt32LE(16, 16);
    arrayBuffer.writeUInt16LE(1, 20);
    arrayBuffer.writeUInt16LE(numberOfChannels, 22);
    arrayBuffer.writeUInt32LE(this.sampleRate, 24);
    arrayBuffer.writeUInt32LE(this.sampleRate * numberOfChannels * 2, 28);
    arrayBuffer.writeUInt16LE(numberOfChannels * 2, 32);
    arrayBuffer.writeUInt16LE(16, 34);
    arrayBuffer.write('data', 36);
    arrayBuffer.writeUInt32LE(length, 40);

    // Данные
    let offset = 44;
    const samples = stereo ? buffer.left.length : buffer.length;

    for (let i = 0; i < samples; i++) {
      if (stereo) {
        const sampleL = Math.max(-1, Math.min(1, buffer.left[i]));
        const sampleR = Math.max(-1, Math.min(1, buffer.right[i]));
        arrayBuffer.writeInt16LE(sampleL < 0 ? sampleL * 0x8000 : sampleL * 0x7FFF, offset);
        arrayBuffer.writeInt16LE(sampleR < 0 ? sampleR * 0x8000 : sampleR * 0x7FFF, offset + 2);
        offset += 4;
      } else {
        const sample = Math.max(-1, Math.min(1, buffer[i]));
        arrayBuffer.writeInt16LE(sample < 0 ? sample * 0x8000 : sample * 0x7FFF, offset);
        offset += 2;
      }
    }

    return arrayBuffer;
  }
}

// Основная функция
async function main() {
  console.log('🎵 Генерация звуковых файлов для Cars RNG...\n');

  const generator = new SimpleAudioGenerator();
  const audioDir = path.join(__dirname, 'public', 'assets', 'audio');
  const musicDir = path.join(audioDir, 'music');
  const soundsDir = path.join(audioDir, 'sounds');

  // Создание директорий
  [audioDir, musicDir, soundsDir].forEach(dir => {
    if (!fs.existsSync(dir)) {
      fs.mkdirSync(dir, { recursive: true });
    }
  });

  // Генерация звуков
  const sounds = {
    'sounds/click.wav': generator.toWav(generator.generateClick()),
    'sounds/button.wav': generator.toWav(generator.generateButton()),
    'sounds/spin.wav': generator.toWav(generator.generateSpin()),
    'sounds/win.wav': generator.toWav(generator.generateWin()),
    'sounds/case-open.wav': generator.toWav(generator.generateCaseOpen()),
    'music/background-music.wav': generator.toWav(generator.generateBackgroundMusic(), true),
  };

  // Сохранение файлов
  for (const [filename, data] of Object.entries(sounds)) {
    const filepath = path.join(audioDir, filename);
    fs.writeFileSync(filepath, data);
    console.log(`✅ Создан: ${filename}`);
  }

  console.log('\n🎉 Все звуки успешно сгенерированы!');
  console.log(`📁 Папка: ${audioDir}`);
}

main().catch(err => {
  console.error('❌ Ошибка:', err);
  process.exit(1);
});
